Python

shaolang opened the pull request:
Written in Python 3.8; uses standard lib's `json` and `copy`.

viebel commented on challenges/python/add_information.py:
```diff
@@ -0,0 +1,20 @@
+import copy
+
+def block_member(library_data, email):
+    data = copy.deepcopy(library_data)
```
There is a way to avoid deep copy. It's called structural sharing. Here is an [implementation](https://blog.klipse.tech/javascript/2021/02/26/structural-sharing-in-javascript.html) with 7 lines of JavaScript. Would you like to port it to Python?

viebel commented on challenges/python/merge_information.py:
```diff
@@ -0,0 +1,63 @@
+    m3 = copy.deepcopy(m1)
+
+    for k, v in m2.items():
+        m3[k] = v
```
It works fine with non-nested maps. But with nested maps it's gonna break.
Is there an implementation of deep merge in Python?

viebel commented on challenges/python/rename_keys.py:
```diff
@@ -0,0 +1,34 @@
+import copy
+
+def rename_keys(m, key_map):
+    return {key_to_use(key_map, k): v for (k, v) in m.items()}
```
Very nice implementation!

viebel commented on challenges/python/retrieve_information.py:
```diff
@@ -0,0 +1,8 @@
+def get_book_property(library_data, isbn, field_name):
+    return library_data['catalog']['booksByIsbn'][isbn][field_name]
```
Could you try to implement a method on hash map that receives a path and return the value at that path, similar to `_.get()` from Lodash?

shaolang commented:
Hihi, I've updated the implementations as requested. I mistaken the challenge is to "implement them however I want to," instead of "implementing them similarly to how the JavaScript version are." LOL

I also didn't know `merge` needs to support deep merge. I've added an example to the script to demonstrate the deep merge.

viebel commented on challenges/python/add_information.py:
```diff
@@ -1,10 +1,14 @@
-    return data
+def __set_in(m, ks, v):
+    updated_m = m.copy()
+    updated_m[ks[0]] = v if len(ks) == 1 else __set_in(m[ks[0]], ks[1:], v)
```
Very nice!

viebel commented on challenges/python/merge_information.py:
```diff
@@ -1,11 +1,18 @@
+        if k not in m3:
+            m3[k] = v
+        else:
+            m3[k] = merge(m3[k], m2[k])
```
Cool!
